javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala

```diff
@@ -0,0 +1,152 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.lang.dec2flt
+
+import java.math.BigInteger
+import java.lang.{Long => jlLong}
+
+/** DIY floating point number with 64-bit significand bits */
+private[dec2flt] class FloatingPoint private (val f: Long, val e: Int) {
+  import FloatingPoint._
+
+  override def toString(): String =
+    s"${this.f} * 2^${this.e}"
+
+  /** Returns a normalized product */
+  def mul(other: FloatingPoint): FloatingPoint = {
+    val low = f * other.f
+    val high = Math.unsignedMultiplyHigh(f, other.f)
+
+    val nlz = jlLong.numberOfLeadingZeros(high)
+    val excessBits = 64 - nlz
+    val shiftedF =
+      if (nlz == 0) high
+      else (high << nlz) | (low >>> excessBits)
+    val newE = (this.e + other.e + excessBits)
+
+    val remainder = low & (-1L >>> nlz)
+    val halfway = 1L << (excessBits - 1)
+
+    val newF = {
+      if (jlLong.compareUnsigned(remainder, halfway) > 0) // roundup
+        shiftedF + 1
+      else if (jlLong.compareUnsigned(remainder, halfway) < 0)
```

Comment:
and
```suggestion
      else if (jlLong.unsigned_<(remainder, halfway))
```
which are faster.